LiteCore/Database/DatabaseImpl.cc

```diff
@@ -625,12 +629,25 @@ namespace litecore {
     // This implements both the public getCollection() and createCollection()
     C4Collection* DatabaseImpl::getOrCreateCollection(CollectionSpec spec, bool canCreate) {
         // Is there already a C4Collection object for it in _collections?
-        LOCK(_collectionsMutex);
-        if (auto i = _collections.find(spec); i != _collections.end())
-            return i->second;                                         // -> Existing object
-
-        // Validate the name (throws if invalid):
+        LOCK(_collectionsMutex);// Validate the name (throws if invalid):
         string keyStoreName = collectionNameToKeyStoreName(spec);
+        CollectionImpl* collectionPtr = nullptr;
+        if (auto i = _collections.find(spec); i != _collections.end()) { 
+            if(!i->second->isValid()) {
+                if(!canCreate) {
+                    if(!_dataFile->keyStoreExists(keyStoreName)) {
+                        // This is an invalid collection, and it has not been recreated elsewhere.
+                        // Furthermore, canCreate is false so don't return a collection object
+                        return nullptr;
+                    } 
+                }
+
+                // After some more validation, we can reset valid to true later
+                collectionPtr = asInternal(i->second.get());
```

Comment:
I would like to do it this way but the issue is that removing from the collection object has the strong potential to immediately free all the memory inside of it, since we don't expect the caller to retain the result of a collection.  That means that the memory inside becomes invalid and turns into a bomb.  If we keep the object alive for at least as long as the database, then that issue is avoided and we only need to make sure that sensible errors are returned.  At least that is what I am told.  @snej Could you elaborate a little more on the race you found here?

Reply:
But there is an asymmetry: we still remove the collection object from _collections of the db instance (this db instance) from which the collection is deleted, c.f. DatabaseImpl::deleteCollection, we only keep it in other db instance, c.f. DatabaseImpl::collectionRemoved. If this is a bump, it could blow in this db instance. Right?

Reply:
Yes, we do this because if you are using the object on one thread like you are supposed to, you will not run into any races.  The race comes because sending this signal to other database instances is a violation of the threading model.